server/notification-providers/nostr.js

```diff
@@ -0,0 +1,108 @@
+const NotificationProvider = require("./notification-provider");
+global.crypto = require("crypto");
+
+const {
+    relayInit,
+    getPublicKey,
+    getEventHash,
+    signEvent,
+    nip04,
+    nip19
+} = require("nostr-tools");
+require("websocket-polyfill");
+
+class Nostr extends NotificationProvider {
+
+    name = "nostr";
+
+    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+
+        // All DMs should have same timestamp
+        const createdAt = Math.floor(Date.now() / 1000);
+
+        const senderPrivateKey = await this.getPrivateKey(notification.sender);
+        const senderPublicKey = getPublicKey(senderPrivateKey);
+        const recipientsPublicKeys = await this.getPublicKeys(notification.recipients);
+
+        if (heartbeatJSON == null) {
+            msg = `🤙💜 Test Notification for [${notification.name}]`;
+        }
+
+        // Create NIP-04 encrypted direct message event for each recipient
+        const events = [];
+        for (const recipientPublicKey of recipientsPublicKeys) {
+            const ciphertext = await nip04.encrypt(senderPrivateKey, recipientPublicKey, msg);
+            let event = {
+                kind: 4,
+                pubkey: senderPublicKey,
+                created_at: createdAt,
+                tags: [[ "p", recipientPublicKey ]],
+                content: ciphertext,
+            };
+            event.id = getEventHash(event);
+            event.sig = signEvent(event, senderPrivateKey);
+            events.push(event);
+        }
+
+        // Publish events to each relay
+        const relays = notification.relays.split("\n");
+        const relayCount = relays.length;
+        let successfulRelays = 0;
+
+        // Connect to each relay
+        for (const relayUrl of relays) {
+            const relay = relayInit(relayUrl);
+            try {
+                await relay.connect();
+                successfulRelays++;
+
+                // Publish events
+                for (const event of events) {
+                    relay.publish(event);
+                }
+            } catch (error) {
+                continue;
+            } finally {
+                relay.close();
+            }
+        }
+
+        // Report success or failure
+        if (successfulRelays > 0) {
+            return `${successfulRelays}/${relayCount} relays connected.`;
+        } else {
+            throw Error("Failed to connect to any relays.");
+        }
```

Comment:
```suggestion
        if (successfulRelays > 0) {
            throw Error("Failed to connect to any relays.");
        }
        return `${successfulRelays}/${relays.length} relays connected.`;
```

Reply:
Thanks, that is simpler but it should now check for succesfulRelays == 0 instead of > 0.

Reply:
True, I updated my comment